tests/e2e/performance/test_performance_testing.py

```diff
@@ -1,55 +1,504 @@
-from os import path
+import multiprocessing
+import shutil
+import sys
+from time import time
 from pathlib import Path
 from string import Template
-from subprocess import run
+from contextlib import contextmanager
+from typing import List, Optional, Tuple, Generator
 
 import pytest
+from starkware.starknet.services.api.contract_class import ContractClass
 
-from tests.e2e.conftest import ACTUAL_CWD
+from protostar.commands.test.test_cases import PassedTestCase
+from protostar.commands.test.test_collector import TestCollector
+from protostar.commands.test.test_runner import TestRunner
+from protostar.commands.test.test_shared_tests_state import SharedTestsState
+from protostar.commands.test.test_suite import TestSuite
+from protostar.utils.compiler.pass_managers import ProtostarPassMangerFactory
+from protostar.utils.starknet_compilation import StarknetCompiler, CompilerConfig
 
-TEST_SUITE_HEADER = """
-%lang starknet
+SCRIPT_DIRECTORY = Path(__file__).parent
 
-from starkware.cairo.common.cairo_builtins import HashBuiltin
 
-@storage_var
-func target() -> (res : felt):
-end
+def _multiply_cases(test_body: str) -> Tuple[str, List[str]]:
+    source_code = ""
+    cases_names = []
+    for i in range(15):
+        case_name = f"test_case_{i}"
+        source_code += Template(
+            """
+            @external
+            func $case_name{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}():
+                $body
+                return ()
+            end
+            """
+        ).substitute(case_name=case_name, body=test_body)
+        cases_names.append(case_name)
+    return source_code, cases_names
 
-"""
 
-TEST_CASE_TEMPLATE = Template(
+def make_test_file(test_body: str, imports: Optional[str] = "") -> Tuple[str, List[str]]:
+    source, cases = _multiply_cases(test_body)
+    return (
+        f"""
+        %lang starknet
+        from starkware.cairo.common.cairo_builtins import HashBuiltin
+        {imports}
+        {source}
+    """,
+        cases,
+    )
+
+
+@pytest.fixture(name="tmp_test_dir")
+def make_test_dir() -> Generator[Path, None, None]:
+    test_dir_path = SCRIPT_DIRECTORY / "tmp_test_files"
+    test_dir_path.mkdir()
+    yield test_dir_path
+    shutil.rmtree(test_dir_path)
+
+
+@pytest.fixture(name="basic_contract_path")
+def get_basic_contract_path(request) -> Path:
+    return (Path(request.node.fspath).parent / "basic.cairo").absolute()
+
+
+@pytest.fixture(name="protostar_cairo_path")
+def get_protostar_cairo_path(request) -> Path:
+    return (Path(request.node.fspath).parent.parent.parent.parent / "cairo").absolute()
+
+
+@contextmanager
+def timing(expected, minimum=None):
+    start_timer = time()
```

Comment:
https://docs.python.org/3/library/time.html I think this might be more precise thing to use for benchmarking

Reply:
I actually think that the code is not that fast anyway, so timer resolution isn't critical here

Reply:
Right.